core/trino-main/src/main/java/io/trino/execution/MemoryRevokingScheduler.java

```diff
@@ -187,26 +189,29 @@ private void scheduleRevoking()
     private synchronized void runMemoryRevoking()
     {
         if (checkPending.getAndSet(false)) {
-            Collection<SqlTask> sqlTasks = null;
+            Collection<SqlTask> allTasks = null;
             for (MemoryPool memoryPool : memoryPools) {
                 if (!memoryRevokingNeeded(memoryPool)) {
                     continue;
                 }
 
-                if (sqlTasks == null) {
-                    sqlTasks = requireNonNull(currentTasksSupplier.get());
+                if (allTasks == null) {
+                    allTasks = requireNonNull(currentTasksSupplier.get());
                 }
 
-                requestMemoryRevoking(memoryPool, sqlTasks);
+                requestMemoryRevoking(memoryPool, allTasks);
             }
         }
     }
 
-    private void requestMemoryRevoking(MemoryPool memoryPool, Collection<SqlTask> sqlTasks)
+    private void requestMemoryRevoking(MemoryPool memoryPool, Collection<SqlTask> allTasks)
     {
         long remainingBytesToRevoke = (long) (-memoryPool.getFreeBytes() + (memoryPool.getMaxBytes() * (1.0 - memoryRevokingTarget)));
-        remainingBytesToRevoke -= getMemoryAlreadyBeingRevoked(sqlTasks, memoryPool);
-        requestRevoking(memoryPool, sqlTasks, remainingBytesToRevoke);
+        List<SqlTask> runningTasksInPool = findRunningTasksInMemoryPool(allTasks, memoryPool);
+        remainingBytesToRevoke -= getMemoryAlreadyBeingRevoked(runningTasksInPool, remainingBytesToRevoke);
+        if (remainingBytesToRevoke > 0) {
```

Comment:
Just curious, is a negative value legitimate, or are you really protecting against calls when the value is zero?

Reply:
I think it's legitimate and it wouldn't necessarily cause problems to pass a negative here, this is just an early return without traversing the task tree if no revoking is going to occur. How this might happen: 
- memory revoking is triggered and initiates a bunch of memory revoking
- more memory is allocated triggering another attempted revoking pass
- the memory in the process of still being revoked is larger than the delta between the current pool's memory and the revoking target